pkg/query/intersection.go

```diff
@@ -22,31 +21,104 @@ func (i *Intersection) addSubIterator(subIt Iterator) {
 }
 
 func (i *Intersection) CheckImpl(ctx *Context, resources []Object, subject ObjectAndRelation) (PathSeq, error) {
+	ctx.TraceEnterIterator("Intersection", resources, subject)
+
 	validResources := resources
 
-	var paths []*Path
+	// Track paths by resource key for combining with AND logic
+	pathsByKey := make(map[string]*Path)
+
+	for iterIdx, it := range i.subIts {
+		ctx.TraceStep("Intersection", "processing sub-iterator %d with %d resources", iterIdx, len(validResources))
 
-	for _, it := range i.subIts {
 		pathSeq, err := it.CheckImpl(ctx, validResources, subject)
 		if err != nil {
 			return nil, err
 		}
-		paths, err = CollectAll(pathSeq)
+		paths, err := CollectAll(pathSeq)
 		if err != nil {
 			return nil, err
 		}
 
+		ctx.TraceStep("Intersection", "sub-iterator %d returned %d paths", iterIdx, len(paths))
+
 		if len(paths) == 0 {
+			ctx.TraceStep("Intersection", "sub-iterator %d returned empty, short-circuiting", iterIdx)
+			var emptyPaths []*Path
+			ctx.TraceExitIterator("Intersection", emptyPaths)
 			return func(yield func(*Path, error) bool) {}, nil
 		}
 
-		validResources = slicez.Map(paths, func(p *Path) Object {
-			return p.Resource
-		})
+		if iterIdx == 0 {
+			// First iterator - initialize pathsByKey using endpoint-based keys
+			for _, path := range paths {
+				key := path.Resource.ObjectType + ":" + path.Resource.ObjectID
+				if existing, exists := pathsByKey[key]; !exists {
+					pathsByKey[key] = path
+				} else {
+					// If multiple paths for same endpoint in first iterator, merge with OR
+					if err := existing.MergeOr(path); err != nil {
+						return nil, err
+					}
+				}
```

Comment:
This logic is handling when you have multiple distinct paths between a resource and a subject that may have different caveat contexts? Is that why we're talking about merges here?